crowdgezwitscher/twitter/models.py

```diff
@@ -173,9 +192,25 @@ def fetch_tweets(self):
         new_tweets.reverse()
 
         for tweet in new_tweets:
-            tweet.save()
-            tweet.hashtags.add(*(tweet_hashtag_mappings.get(tweet.tweet_id, [])))
-            tweet.save()
+            try:
+                tweet.save()
+            except IntegrityError as e:
+                # This should not have happened.
+                # Assumption: We have an inconsistent state where an account's last_known_tweet_id has not been updated
+                # after saving a tweet. The reason might be a crash or the server shutting down or whatever.
+                # So the DB now complains about saving a new tweet with a non-unique tweet_id.
+                # To fix this we reset the account's last_known_tweet_id to its latest known tweet's ID. As new_tweets
+                # has been fetched with a wrong last_known_tweet_id, we stop saving more of them and will try again the
+                # next time this method is called.
+                logger.warning(
+                    "Failed to save tweet for account #%i. Setting last_known_tweet_id to latest tweet's ID. "
+                    "Exception: %s", self.id, e
+                )
+                last_known_tweet_id = self.tweet_set.latest('pk').tweet_id
```

Comment:
Can we really assume that the tweet with the latest `pk` has the highest `tweet_id`? This would mean that tweets are always received in order, I have no idea if this is given. In case we are not sure, we could filter for the highes `tweet_id`?

Reply:
Our whole algorithm in `_fetch_tweets` relies on this assumption :)

Reply:
Okay. I am really confident that we quadruple-checked that while implementing it :)